### PR TITLE
Optional gate gradient identity for the grouped MoE combine

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

In the grouped MoE combine, differentiating `out = down(inter) * gate` pins the down-projection output on the autograd tape solely to form the routing-weight gradient `dGate = <dOut, Y>`. That tensor is large at long sequence lengths.

## What this does

With `UNSLOTH_MOE_GATEGRAD=1` (default off), `forward_native_grouped_mm` derives the gate gradient from the inner-product identity `dGate = <A, dA> / gate` over the pre-down-projection activation instead, so Y never has to stay on the tape. The identity is exact for any linear down projection (frozen base, the recompute path, additive LoRA) but not for a bias added after the down matmul, so the path auto-disables when a down-projection bias is present.

## Testing

- 4 CPU unit tests check the identity against plain autograd in fp32, with and without a LoRA term on the down weight, plus forward-identity and env gating.
- Full-model loss parity on Qwen3-30B-A3B: max per-step deviation 0.0063 over 20 deterministic steps.
- Throughput neutral; peak memory down 1.41 GiB (6 percent) at 16k tokens, with the saving growing with sequence length.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Fixed the backward pass: it returns early when the incoming gradient is None (frozen down projection) and only computes the gate gradient when it is actually required. The gate divide now uses a signed floor so a negative routing weight keeps its sign instead of flipping (gemma4_moe folds an unconstrained scale into the weight, so gates can be negative). Also switched the two `top_k_weights.view(-1)` calls to `.reshape(-1)` for non-contiguous inputs. Added tests for the negative-gate and frozen-router cases.

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 1]
```diff
@@ -1171,6 +1204,17 @@ def forward_native_grouped_mm(
     else:
         inter = F.silu(gate) * up
 
+    # Gate gradient via the <A, dA> identity (env-gated) so the combine never
+    # caches Y just for the gate multiply. Disabled when a down bias exists.
+    _gategrad = (
+        _moe_gategrad_enabled()
+        and getattr(self, "down_proj_bias", None) is None
+    )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py:1441]: Good catch. The identity is now skipped when top_k_weights does not require grad, so nothing extra is kept alive for a frozen router.

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 2]
```diff
@@ -1244,9 +1288,13 @@ def forward_native_grouped_mm(
         raise AttributeError("MoE layer must have 'down_proj' or 'w2'.")
 
     # Apply routing weights and scatter-add (reduce).
-    flat_weights = top_k_weights.view(-1)
-    permuted_weights = flat_weights[sorted_indices]
-    mm2_out = mm2_out * permuted_weights.unsqueeze(-1)
+    if _gategrad:
+        # Gate grad comes from the identity; detach so the multiply does not pin Y.
+        mm2_out = mm2_out * permuted_weights.detach().unsqueeze(-1)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py:1542]: Marked the backward once_differentiable, so create_graph now raises instead of silently returning wrong second derivatives. Added a test.

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 3]
```diff
@@ -1001,6 +1001,39 @@ def patch_param_wrapper_for_moe():
...
+    @staticmethod
+    def forward(ctx, inter, permuted_weights):
+        ctx.save_for_backward(inter, permuted_weights)
+        return inter
+
+    @staticmethod
+    def backward(ctx, grad_inter):
+        inter, gate = ctx.saved_tensors
+        dgate = (inter.to(torch.float32) * grad_inter.to(torch.float32)).sum(dim=-1)
+        dgate = dgate / gate.to(torch.float32).clamp_min(1e-12)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py]: An exact zero selected weight carries no recoverable signal in this formulation since the detached multiply zeroes the incoming gradient, so it degrades to a zero gradient rather than NaN. Detecting zeros at runtime would add a device sync per layer, and the affected archs route fp16 requests through forced float32 mixed precision, so keeping the documented behavior.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 4]
```diff
@@ -1001,6 +1001,61 @@ def patch_param_wrapper_for_moe():
...
+            # Divide by the signed routing weight: a negative gate (e.g. Gemma4's
+            # per_expert_scale folded into top_k_weights) must keep its sign, so
+            # clamping toward +eps would flip it. Floor the magnitude to avoid
+            # dividing by ~0; a gate that underflows to zero carries no
+            # recoverable gate gradient, so it collapses to 0 rather than NaN.
+            gate_f = gate.to(torch.float32)
+            safe_gate = torch.where(
+                gate_f >= 0, gate_f.clamp_min(1e-12), gate_f.clamp_max(-1e-12)
+            )
+            grad_gate = (dgate / safe_gate).to(gate.dtype)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py:1249]: Fixed: the call site now sign-floors the gate away from zero (straight-through) and the multiply and the divide share that same floored value, so the floor cancels exactly and the true gradient survives a zero routing weight. Added a zero-gate test against autograd.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 5]
```diff
@@ -1244,9 +1327,13 @@ def forward_native_grouped_mm(
         raise AttributeError("MoE layer must have 'down_proj' or 'w2'.")
 
     # Apply routing weights and scatter-add (reduce).
-    flat_weights = top_k_weights.view(-1)
-    permuted_weights = flat_weights[sorted_indices]
-    mm2_out = mm2_out * permuted_weights.unsqueeze(-1)
+    if _gategrad:
+        # Gate grad comes from the identity; detach so the multiply does not pin Y.
+        mm2_out = mm2_out * permuted_weights.detach().unsqueeze(-1)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py:1542]: Good catch. The floor is now computed in float32, so eps stays 1e-12 even for fp16 routing weights: the masked-route leak is bounded by 1e-12 * |Y|, below any dtype resolution, and the same-value cancellation still gives the exact gate gradient. Added an fp16 zero-gate test.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 6]
```diff
@@ -1171,6 +1228,36 @@ def forward_native_grouped_mm(
...
+        # gradient still reaches top_k_weights). The multiply below and the
+        # identity's divide use the SAME floored value, so they cancel exactly and
+        # the gate gradient <Y, dOut> survives even a routing weight of exactly
+        # zero. Floor in float32 so eps stays 1e-12 even for fp16 routing weights
+        # (fp16's smallest normal is ~6e-5, large enough to visibly leak masked
+        # routes into the output); the forward then changes by at most
+        # 1e-12 * |Y| for |gate| < 1e-12, far below any dtype's resolution.
+        if raw_weights.dtype == torch.float16:
+            raw_weights = raw_weights.to(torch.float32)
+        eps = 1e-12
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py:1455]: Correct, and distinct from the earlier gate upcast: with fp16 activations the incoming grad_inter carries the tiny floored gate as a factor and underflows fp16 to zero before backward, so the zero-route router gradient is lost (bf16 and fp32 hold it, verified). fp16 now falls back to the standard multiply, which keeps the exact <Y, dOut> gradient; added a regression test.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 7]
```diff
@@ -1001,6 +1001,39 @@ def patch_param_wrapper_for_moe():
...
+    def forward(ctx, inter, permuted_weights):
+        ctx.save_for_backward(inter, permuted_weights)
+        return inter
+
+    @staticmethod
+    def backward(ctx, grad_inter):
+        inter, gate = ctx.saved_tensors
+        dgate = (inter.to(torch.float32) * grad_inter.to(torch.float32)).sum(dim=-1)
+        dgate = dgate / gate.to(torch.float32).clamp_min(1e-12)
+        return grad_inter, dgate.to(gate.dtype)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py]: Backward now returns early when grad_inter is None (frozen down projection). The call site also sign-floors the gate, and the multiply and the divide share that same floored value, so a tiny gate keeps its exact gradient instead of being shrunk by the clamp.

File: unsloth_zoo/temporary_patches/moe_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py]: The divide uses a signed floor now: clamp_min for positive weights, clamp_max for negative. A negative folded per-expert scale keeps its sign, and the multiply and divide share that floored value so the identity still holds.

File: unsloth_zoo/temporary_patches/moe_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py]: Added both. Backward returns early on a None grad_inter and only computes the gate gradient when needs_input_grad[1] is set, so a frozen router does no extra work.

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 8]
```diff
@@ -1171,6 +1204,17 @@ def forward_native_grouped_mm(
...
 
+    # Gate gradient via the <A, dA> identity (env-gated) so the combine never
+    # caches Y just for the gate multiply. Disabled when a down bias exists.
+    _gategrad = (
+        _moe_gategrad_enabled()
+        and getattr(self, "down_proj_bias", None) is None
+    )
+    permuted_weights = None
+    if _gategrad:
+        permuted_weights = top_k_weights.view(-1)[sorted_indices]
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py]: Both top_k_weights.view(-1) calls are reshape(-1) now.

**Comment by danielhanchen** [comment]
### Final results

Setup: optional gate-gradient identity for the grouped MoE combine (`UNSLOTH_MOE_GATEGRAD=1`, default off), Qwen3-30B-A3B, 16k tokens.

| metric | result |
|---|---|
| peak memory saved at 16k | -1.41 GiB (6%) |
| throughput | neutral |
| loss parity (max per-step deviation, 20 deterministic steps) | 0.0063 |

Takeaway: deriving the gate gradient from the inner-product identity keeps the down-projection output off the autograd tape, saving about 6% peak memory at 16k tokens with the saving growing at longer sequence lengths, and throughput stays neutral.

**Comment by danielhanchen** [comment]
@codex review
